icu4c/source/common/uloc.cpp

```diff
@@ -1184,104 +1184,128 @@ _getLanguage(const char* localeID,
+        int32_t offset = _findIndex(LANGUAGES_3, buffer);
         if(offset>=0) {
-            result.clear();
-            result.append(LANGUAGES[offset], status);
+            const char* const alias = LANGUAGES[offset];
+            sink->Append(alias, (int32_t)uprv_strlen(alias));
+            return;
         }
     }
 
-    if (sink != nullptr && !result.isEmpty()) {
-        sink->Append(result.data(), result.length());
-    }
-
-    *pEnd = localeID;
+    sink->Append(buffer, len);
 }
 
 static void
 _getScript(const char* localeID,
            ByteSink* sink,
-           const char** pEnd,
-           UErrorCode& status) {
+           const char** pEnd) {
     U_ASSERT(pEnd != nullptr);
-    CharString result;
-    int32_t idLen = 0;
-
     *pEnd = localeID;
 
-    /* copy the second item as far as possible and count its length */
-    while(!_isTerminator(localeID[idLen]) && !_isIDSeparator(localeID[idLen])
-            && uprv_isASCIILetter(localeID[idLen])) {
-        idLen++;
-    }
+    constexpr int32_t LENGTH = 4;
 
-    /* If it's exactly 4 characters long, then it's a script and not a country. */
-    if (idLen == 4) {
-        int32_t i;
-        *pEnd = localeID + idLen;
-        if (idLen >= 1) {
-            result.append((char)uprv_toupper(*(localeID++)), status);
-        }
-        for (i = 1; i < idLen; i++) {
-            result.append((char)uprv_tolower(*(localeID++)), status);
-        }
+    int32_t len = 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len]) &&
+            uprv_isASCIILetter(localeID[len])) {
+        if (len == LENGTH) { return; }
+        len++;
     }
+    if (len != LENGTH) { return; }
+
+    *pEnd = localeID + LENGTH;
+    if (sink == nullptr) { return; }
 
-    if (sink != nullptr && !result.isEmpty()) {
-        sink->Append(result.data(), result.length());
+    char scratch[LENGTH];
+    int32_t capacity = 0;
+    char* buffer = sink->GetAppendBuffer(
+            LENGTH, LENGTH, scratch, UPRV_LENGTHOF(scratch), &capacity);
+
+    buffer[0] = uprv_toupper(localeID[0]);
+    for (int32_t i = 1; i < LENGTH; ++i) {
+        buffer[i] = uprv_tolower(localeID[i]);
     }
+
+    sink->Append(buffer, LENGTH);
 }
 
 static void
 _getRegion(const char* localeID,
            ByteSink* sink,
-           const char** pEnd,
-           UErrorCode& status) {
+           const char** pEnd) {
     U_ASSERT(pEnd != nullptr);
-    CharString result;
-    int32_t idLen=0;
+    *pEnd = localeID;
+
+    constexpr int32_t MINLEN = 2;
+    constexpr int32_t MAXLEN = ULOC_COUNTRY_CAPACITY - 1;  // Minus NUL.
 
-    /* copy the country as far as possible and count its length */
-    while(!_isTerminator(localeID[idLen]) && !_isIDSeparator(localeID[idLen])) {
-        result.append((char)uprv_toupper(localeID[idLen]), status);
-        idLen++;
+    int32_t len = 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len])) {
+        if (len == MAXLEN) { return; }
```

Comment:
optional / also not a new problem: we seem to allow things other than letters and digits in region subtags